### PR TITLE
implements transport efficiency in flow accumulation (#70) 

### PR DESCRIPTION
**Pull request by itati01**
As suggested with separate loops to avoid performance hits.

### CODE REVIEW

**Comment by mdbartos** [comment]
Hi @itati01 

Let me know when you're ready to submit and I will review.

Thanks!
MDB

**Comment by itati01** [comment]
Hi @mdbartos ,
Sorry, I am still not familiar with git. Are my commits / pull requests not available to you?

**Comment by mdbartos** [comment]
Hi @itati01 

I can see your changes. I just wanted to make sure you had everything ready to go before I reviewed and pulled the changes in (there were a couple commented out lines).

**Comment by mdbartos** [comment]
A couple things:

1) It looks like you fixed #68 in this PR. Can you go ahead and fix issue #74 and include it in this PR as well?
2) Can you add a test for accumulation with efficiency to `tests/test_grid.py`? The test that you referenced in #70 should be fine.

**Comment by itati01** [comment]
Ok, will fix #74 and have look into test_grid.py.

Feel free to remove any comment. I was unsure about the `flatten()` in `efficiency.flatten() / 100.` to convert percentages to relative values. Here I simply followed your steps with the weight raster.

**Comment by mdbartos** [comment]
Hi @itati01 

I believe using flatten to create a flat copy is the correct approach (otherwise you can end up modifying the original array inside the function). For example:

```python
import numpy as np

def test_modify(a):
    a[0] = 0
    
def test_ravel(a):
    a = a.ravel()
    a[0] = 0
    
def test_flatten(a):
    a = a.flatten()
    a[0] = 0
```

```python
a = np.ones(3, dtype=int)
test_modify(a)
print(a)
>>> [0 1 1]

a = np.ones(3, dtype=int)
test_ravel(a)
print(a)
>>> [0 1 1]

a = np.ones(3, dtype=int)
test_flatten(a)
print(a)
>>> [1 1 1]
```

**Comment by itati01** [comment]
Hi @mdbartos,

fixed #74, created a new GeoTIFF, extended test_grid.py to cover a basic testing of the transport efficiency, and minor issues with transport efficiency. So, the PR should be ready.

**Comment by mdbartos** [comment]
Looks like coveralls is complaining because there's no test for efficiency with dinf accumulation. Might be worth adding to the tests.

**Comment by itati01** [comment]
This would be `grid.accumulation(data='dinf_dir', dirmap=dirmap, out_name='dinf_acc', routing='dinf', efficiency=grid.eff)` in test_accumulation but I have to figure out the correct solution first.

However, there is a problem which should also affect the (still untested) weighting: `grid.clip_to('patch')` does not apply to the efficiency raster. Shall I create and use an unclipped dinf flow direction to test this function?

**Comment by mdbartos** [comment]
> This would be `grid.accumulation(data='dinf_dir', dirmap=dirmap, out_name='dinf_acc', routing='dinf', efficiency=grid.eff)` in test_accumulation but I have to figure out the correct solution first.
> 
> However, there is a problem which should also affect the (still untested) weighting: `grid.clip_to('patch')` does not apply to the efficiency raster. Shall I create and use an unclipped dinf flow direction to test this function?

I think you can get around this by using `clip_to('dir')` to return the view to the original raster dimensions. As a side note, I want to expand the `clip_to` function so that it takes any `Raster` as an input (instead of just named datasets). See #65 e.g.

**Comment by itati01** [comment]
I addressed the clipping issue by using `_input_handler()` for weights and transport efficiency in `accumulation`, similar to flow direction. So, we can new pass their names. Flattening is still required, though. Shall I open an issue? Going to remove the percent parts.

File: pysheds/grid.py [hunk 1]
```diff
@@ -1123,6 +1123,9 @@ def accumulation(self, data, weights=None, dirmap=None, nodata_in=None, nodata_o
                  List of integer values representing the following
                  cardinal and intercardinal directions (in order):
                  [N, NE, E, SE, S, SW, W, NW]
+        efficiency: numpy ndarray
+                 Array of relative efficiency factors (0..1) or percentages (0..100)
```
[the remark this answers is not in the document]
Comment (itati01) [reply pysheds/grid.py]: see new commit

File: pysheds/grid.py [hunk 2]
```diff
@@ -1203,20 +1206,47 @@ def _d8_accumulation(self, fdir=None, weights=None, dirmap=None, nodata_in=None,
...
                 acc = (~nodata_cells).ravel().astype(int)
-            # TODO: Does this need to have a min length?
+            # efficiency either percentage or relative
+            # percentage converted to relative values
+            if efficiency is not None:
+                assert(efficiency.size == fdir.size)
+                eff = efficiency.flatten() # must be flattened to avoid IndexError below
+                acc = acc.astype(float)
+                eff_max, eff_min = np.max(eff), np.min(eff)
+                if eff_max>1:       # percent?
```
[the remark this answers is not in the document]
Comment (itati01) [reply pysheds/grid.py]: ditto

File: pysheds/grid.py [hunk 3]
```diff
@@ -1282,6 +1312,20 @@ def _dinf_accumulation(self, fdir=None, weights=None, dirmap=None, nodata_in=Non
                 acc = weights.flatten().astype(float)
             else:
                 acc = (~nodata_cells).ravel().astype(float)
+            # efficiency either percentage or relative
+            # percentage converted to relative values
+            if efficiency is not None:
+                assert(efficiency.size == fdir.size)
+                eff = efficiency.flatten()
+                eff_max, eff_min = np.max(eff), np.min(eff)
+                if eff_max>1:       # percent?
```
[the remark this answers is not in the document]
Comment (itati01) [reply pysheds/grid.py]: ditto

**Comment by itati01** [comment]
Have opened an issue regarding dinf accumulation in general (#76). Nonetheless, added a simple test case for the efficiency.

**Comment by mdbartos** [comment]
Hi @itati01 

Thanks for all your hard work! I think it's almost ready to pull in.

> I addressed the clipping issue by using `_input_handler()` for weights and transport efficiency in `accumulation`, similar to flow direction. So, we can new pass their names. Flattening is still required, though. Shall I open an issue?

I like the idea of using `_input_handler()` for the weights/efficiency. It would make things more consistent overall. There are a couple issues though:

1) Wouldn't passing in an ndarray for `weights` now raise an error? That would be a breaking change.
2) This same behavior would be needed for the other methods that use weights (such as `flow_distance`).

It might be best to save this change for later, at least until these two things are resolved.

I think you can get the behavior you want by using `grid.view` to return a view of the efficiency raster at the grid's current dimensions. So, something like:

```python
grid.clip_to('dir')
eff = grid.view('eff')
grid.accumulation(data='dir', dirmap=dirmap, out_name='acc_eff', efficiency=eff)
```

See here for more info: https://mdbartos.github.io/pysheds/views.html

**Comment by itati01** [comment]
Works, thanks. Why is this not working with grid.view()

> grid.eff[grid.eff==grid.eff.nodata] = 1 # change nodata to 1
> grid.clip_to("catch")
> assert(grid.view("eff").min() == grid.eff.min())

but 

> grid.clip_to("dir")
> assert(grid.view("eff").min() == grid.eff.min())

**Comment by mdbartos** [comment]
> Works, thanks. Why is this not working with grid.view()
> 
> > grid.eff[grid.eff==grid.eff.nodata] = 1 # change nodata to 1
> > grid.clip_to("catch")
> > assert(grid.view("eff").min() == grid.eff.min())
> 
> but
> 
> > grid.clip_to("dir")
> > assert(grid.view("eff").min() == grid.eff.min())

I will take a look later today and let you know.

**Pull request merged**